mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerTimezone.java

```diff
@@ -0,0 +1,165 @@
+import org.apache.hadoop.hive.serde2.io.DateWritable;
+import org.apache.hadoop.hive.serde2.io.TimestampWritable;
+import org.apache.iceberg.FileFormat;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.data.Record;
+import org.apache.iceberg.mr.TestHelper;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.types.Types;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+import static org.apache.iceberg.types.Types.NestedField.optional;
+import static org.junit.runners.Parameterized.Parameter;
+import static org.junit.runners.Parameterized.Parameters;
+
+@RunWith(Parameterized.class)
+public class TestHiveIcebergStorageHandlerTimezone {
+
+  @Parameters(name = "timezone={0}")
+  public static Collection<Object[]> parameters() {
+    Collection<Object[]> testParams = ImmutableList.of(
+        new String[] {"America/New_York"},
+        new String[] {"Asia/Kolkata"},
+        new String[] {"UTC/Greenwich"}
+    );
+
+    return testParams;
+  }
+
+  private TestHiveShell shell;
+
+  private TestTables testTables;
+
+  private TimeZone originalTz = TimeZone.getDefault();
+
+  @Parameter(0)
+  public String timezoneString;
+
+  @Rule
+  public TemporaryFolder temp = new TemporaryFolder();
+
+  @Before
+  public void before() throws IOException, IllegalAccessException {
+    TimeZone.setDefault(TimeZone.getTimeZone(timezoneString));
+
+    try {
+      // Magic to update cached stuff for Hive where the default timezone is used
+      Field formatterField = TimestampWritable.class.getDeclaredField("threadLocalDateFormat");
+      formatterField.setAccessible(true);
+      ThreadLocal<DateFormat> formatter = (ThreadLocal<DateFormat>) formatterField.get(null);
+      formatter.set(new SimpleDateFormat("yyyy-MM-dd HH:mm:ss"));
+
+      Field localTimeZoneField = DateWritable.class.getDeclaredField("LOCAL_TIMEZONE");
+      localTimeZoneField.setAccessible(true);
+      ThreadLocal<TimeZone> localTimeZone = (ThreadLocal<TimeZone>) localTimeZoneField.get(null);
+      localTimeZone.set(TimeZone.getDefault());
+    } catch (NoSuchFieldException nse) {
+      // Hive2 has thread local cache we have to clean, so the new default timezone is used when formatting
+      // Hive3 we do not have this for Date and Timestamp
+    }
+
+    this.shell = HiveIcebergStorageHandlerTestUtils.shell();
+    this.testTables = HiveIcebergStorageHandlerTestUtils.testTables(shell, TestTables.TestTableType.HIVE_CATALOG, temp);
+
+    // Uses spark as an engine so we can detect if we unintentionally try to use any execution engines
+    HiveIcebergStorageHandlerTestUtils.init(shell, testTables, temp, "spark");
+  }
+
+  @After
+  public void after() {
+    shell.stop();
+    TimeZone.setDefault(originalTz);
```

Comment:
Should we also be resetting the ThreadLocals here so that other tests which will run in the same JVM will not be affected?

Reply:
Let's talk about this.
I do not think that I am able to fix all of the places that is needed to have a clean test run in the same JVM after this.
The best I could do is fixing the things I know of, and I would avoid this to highlight the issues faster.
Maybe I should even remove the `TimeZone.setDefault(originalTz)` too? I left it there because some logging/reporting might use that

What do you think?

Reply:
I think we should either reset both the ThreadLocals and default timezone or neither. If we just do one of them, further Hive tests dealing with timezones may see a different timezone inside Hive code because of an older ThreadLocal and the Iceberg code may use the current Java timezone causing false positives. Looking at the test results though, doesn't seem like its actually affecting anything.

Reply:
Then I think we agree that we should not try to revert the TZ.

Reply:
Done